base/frequencies.py

```diff
@@ -568,6 +570,162 @@ def test_nested_estimator():
+
+                # Zero out counts from pivots after the given maximum date to
+                # prevent numerical errors associated with normalizing very
+                # small values to 1.
+                if max_date is not None:
+                    counts[pivots > max_date] = 0.0
+
+            freq_matrix[i] = counts
+
+        return freq_matrix
+
+    @classmethod
+    def normalize_frequencies(cls, freq_matrix):
+        """Normalize the values of a given frequency matrix to 1 across all columns with
+        non-zero sums.
+        """
+        normalized_freq_matrix = freq_matrix.copy()
+
+        # Find columns that can be divided by their sum.
+        nonzero_columns = np.nonzero(freq_matrix.sum(axis=0))[0]
+
+        # Normalize by column.
+        normalized_freq_matrix[:, nonzero_columns] = freq_matrix[:, nonzero_columns] / freq_matrix[:, nonzero_columns].sum(axis=0)
+
+        return normalized_freq_matrix
+
+    @classmethod
+    def estimate_frequencies(cls, tip_dates, pivots, tip_populations, **kwargs):
+        """Estimate frequencies of the given observations across the given pivots.
+
+        Initial frequencies are weighted by the size of the population where
+        each observation originated and the resulting weighted values are
+        normalized to sum to 1.
+        """
+        # Calculate base frequencies from observations.
+        freq_matrix = cls.get_frequencies_from_observations(tip_dates, pivots, **kwargs)
+
+        # Weight initial frequencies by region population size.
+        weighted_freq_matrix = freq_matrix * tip_populations
+
+        # Normalize frequencies to sum to 1.
+        normalized_freq_matrix = cls.normalize_frequencies(weighted_freq_matrix)
+
+        return normalized_freq_matrix
+
+    @classmethod
+    def estimate_frequencies_for_tree(cls, tree, pivots, regions, **kwargs):
+        """Estimate global and regional frequencies for all nodes in a tree across the
+        given pivots.
+
+        Frequencies are weighted by the population size of the each individual
+        observation's region as defined in the given regions tuple.
+
+        """
+        tips = np.array(sorted([(tip.clade, tip.attr["num_date"], tip.attr["region"])
+                                for tip in tree.get_terminals()], key=lambda row: row[1]))
+
+        clades = tips[:, 0].astype(int)
+        tip_dates = tips[:, 1].astype(float)
+        tip_regions = tips[:, 2]
+
+        population_by_region = {region_data[0]: region_data[2] for region_data in regions}
+        total_population = 0
+        for region, population in population_by_region.items():
+            total_population += population
+
+        tip_populations = []
+        for region in tip_regions:
+            tip_populations.append(population_by_region[region] / total_population)
+
+        tip_populations = np.array(tip_populations).reshape((len(tip_populations), -1))
```

Comment:
is this the same as np.array([tip_populations])? what shape is this supposed to be in afterwards?